sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/BufferSendState.scala

```diff
@@ -0,0 +1,232 @@
+    }
+    isClosed = true
+    freeBounceBuffers()
+    request.close()
+  }
+
+  case class RangeBuffer(
+      range: BlockRange[SendBlock], rapidsBuffer: RapidsBuffer)
+      extends AutoCloseable {
+    override def close(): Unit = {
+      rapidsBuffer.close()
+    }
+  }
+
+  def next(): AddressLengthTag = synchronized {
+    require(acquiredBuffs.isEmpty,
+      "Called next without calling `releaseAcquiredToCatalog` first")
+
+    var bounceBuffToUse: MemoryBuffer = null
+    var buffOffset = 0L
+
+    val buffsToSend = try {
+      if (!markedDone) {
+        var deviceBuffs = 0L
+        var hostBuffs = 0L
+        acquiredBuffs = blockRanges.safeMap { blockRange =>
+          val transferRequest =
+            blockRange.block.bufferTransferRequest
+
+          // we acquire these buffers now, and keep them until the caller releases them
+          // using `releaseAcquiredToCatalog`
+          closeOnExcept(
+            requestHandler.acquireShuffleBuffer(transferRequest.bufferId())) { rapidsBuffer =>
+            //these are closed later, after we synchronize streams
+            rapidsBuffer.storageTier match {
+              case StorageTier.DEVICE =>
+                deviceBuffs += blockRange.rangeSize()
+              case _ => // host/disk
+                hostBuffs += blockRange.rangeSize()
+            }
+            RangeBuffer(blockRange, rapidsBuffer)
+          }
+        }
+
+        logDebug(s"Occupancy for bounce buffer is [device=${deviceBuffs}, host=${hostBuffs}] Bytes")
+
+        bounceBuffToUse = if (deviceBuffs >= hostBuffs || hostBounceBuffer == null) {
+          deviceBounceBuffer.buffer
+        } else {
+          hostBounceBuffer.buffer
+        }
+
+        acquiredBuffs.foreach { case RangeBuffer(blockRange, rapidsBuffer) =>
+          require(blockRange.rangeSize() <= bounceBuffToUse.getLength - buffOffset)
+          withResource(rapidsBuffer.getMemoryBuffer) { memBuff =>
+            bounceBuffToUse match {
+              case _: HostMemoryBuffer =>
+                //TODO: HostMemoryBuffer needs the same functionality that
+                // DeviceMemoryBuffer has to copy from/to device/host buffers
+                CudaUtil.copy(
+                  memBuff,
+                  blockRange.rangeStart,
+                  bounceBuffToUse,
+                  buffOffset,
+                  blockRange.rangeSize())
+              case d: DeviceMemoryBuffer =>
+                memBuff match {
+                  case mh: HostMemoryBuffer =>
+                    // host original => device bounce
+                    d.copyFromHostBufferAsync(buffOffset, mh, blockRange.rangeStart,
+                      blockRange.rangeSize(), serverStream)
+                  case md: DeviceMemoryBuffer =>
+                    // device original => device bounce
+                    d.copyFromDeviceBufferAsync(buffOffset, md, blockRange.rangeStart,
+                      blockRange.rangeSize(), serverStream)
```

Comment:
Where does the synchronization for these copies occur?  I don't see a sync on the server stream before this completes.  If intentional then this async behavior needs to be well documented, which makes me question whether the iterator API is really an appropriate fit to convey what is happening.

Reply:
Adding a comment for now, but yeah I agree this is a odd. You have to know what `next` is doing.